hw/dv/dpi/usbdpi/usb_transfer.h

```diff
@@ -0,0 +1,146 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_HW_DV_DPI_USBDPI_USB_TRANSFER_H_
+#define OPENTITAN_HW_DV_DPI_USBDPI_USB_TRANSFER_H_
+#include <stdint.h>
+
+// USBDEV IP block supports up to 64-bytes/packet
+#ifndef USBDEV_MAX_PACKET_SIZE
+#define USBDEV_MAX_PACKET_SIZE 64U
+#endif
+
+// Maximal length of a byte sequence to be transmitted/receive without
+// interruption or pause (Start Of Frame, Setup packet, OUT Data packet
+// and End Of Frame all back-to-back?)
+#define USBDPI_MAX_DATA (3U + 3U + 1U + USBDEV_MAX_PACKET_SIZE + 2U + 1U)
+
+/**
+ * Forwards reference to usbpdi state context
+ */
+typedef struct usbdpi_ctx usbdpi_ctx_t;
+
+/**
+ * Description of a transfer over the USB
+ *
+ * A control transfer
+ *    Setup stage - [ Data stage ] - Status stage
+ * A data transfer
+ *    Setup stage - Data stage - [ Status stage ]
+ */
```

Comment:
Can you clarify what is meant by control transfer vs. a "data" transfer? The word "transfer" has a specific meaning in USB, as does the word "setup", and there are specific associations between terms (like how a "stage" generally refers to the phases of a control transfer). Those diagrams both look like control transfers to me.

Actually, as I look at this file and the implementation, are you certain you don't mean "transaction"? It is critical that the terms are not substituted for each other:

`packet`: The smallest identifiable message unit. Boundaries are SOP (always followed by PID) and EOP symbols.
`transaction`: A sequence of up to 3 packets that represents the smallest unit of service. Starts with a token packet and can optionally contain data and/or handshake packets, depending on the endpoint type.
`transfer`: A sequence of transactions that represent a complete message, request, etc. Control transfers have a special, defined sequence, where there are setup, data, and status stages.

Reply:
Good point, and apologies for the misleading diagrams (I was unclear - less clear? - on some of the protocol details when starting to create the usbdpi_transfer_t container).
These are not USB transfers as such, but a description of a transfer to/from the usbdpi model, and its present use is a stepping stone to extending the behavior of the DPI model.
At the moment they are, as you note, transactions in practice, but may in time become full control/data transfers with the different transactions being further apart in time, and interleaving traffic to/from other endpoints.
I think it's simplest to leave the container and functions called usbdpi_transfer_t and transfer_x() for now, if that's okay, but perhaps when its final usage is clearer, an alternative name may be used.

Reply:
I've made the initial changes as requested, to avoid confusion hopefully. No, you don't sound pedantic: we're agreed I think, and I'm trying to make the DPI extensions incrementally. I guess it's not yet clear how far it should be developed before the focus switches more to UVM and DV testing so that we can exercise it more completely. It definitely makes sense to keep the terminology consistent with that testing.

I shall aim to marshal my thoughts and write them up before creating an Issue re further DPI model following the streaming test that I've been building.

At the moment, locally I have a streaming test that passes a LFSR-generated bytestream from the CPU to the DPI model as packets of variable sizes. The DPI model or equivalent Linux-hosted application then XORs this stream with the output of its own LFSR, before streaming the result back to the device, where the CPU checks it. This is passing on CW-310 FPGA and in Verilator t-l sim with the DPI.